### DISCUSSION TITLE
Project migration from VB6 to TB

### DISCUSSION THREAD

**Discussion by wqweto** [twinbasic/twinbasic#550] [General]
Just started test migration to TB of a 16k LOC [project of ours](https://github.com/Unicontsoft/UcsFiscalPrinters) which we use in production here.

After taking care of the few errors and warnings out of incompatibilities now the project won't compile to DLL and does not want to start debug session on running for instance `Sub Main`.

Attaching FYI the workspace as a .twinproj file: [UcsFP20.twinproj.zip](https://github.com/WaynePhillipsEA/twinbasic/files/7693612/UcsFP20.twinproj.zip)

(Still not sure if .twinproj file includes everything for the workspace to be loaded in the IDE.)

**Comment by WaynePhillipsEA** [comment]
Check for message boxes that might be hiding beneath the IDE.  Looks like it's this:
`    Private Declare PtrSafe Function ArrPtr Lib "vbe7" Alias "VarPtr" (Ptr() As Any) As LongPtr`

Building the DLL is crashing though, so I'll get that fixed.

**Reply by wqweto** [reply]
It was not the intention to use `PtrSafe` declares for x86 target. A (dormant) bug on my part as this is compiled in VB6 only but the JSON module is VBA-aware.

Fixed it by checking for `Win64` conditional too.
```
#Const HasPtrSafe = (VBA7 <> 0 And Win64 <> 0)
```
Now debug session starts ok but `Build` button is still not producing any output.

**Reply by WaynePhillipsEA** [reply]
Yeah, the build is borked.   I've fixed it here, just running tests then I'll publish.

**Reply by bclothier** [reply]
> It was not the intention to use `PtrSafe` declares for x86 target. 

Not relevant to the issue but FWIW: the `PtrSafe` can and should be used in both targets, not just for the 64-bit. The only difference is that in 32-bit target, it's optional but in 64-bit target it's mandatory. For most declares, `VBA7` conditional check is usually sufficient to ensure that it's OK to use the keyword `PtrSafe`. 

In VBA world and for new projects, I've stopped writing conditionals for VBA6/VBA7 since it's now out of support and just write declares with `PtrSafe` as the default regardless whether it's 32-bit or 64-bit.

**Reply by WaynePhillipsEA** [reply]
OK, that should now build with v0.13.29, however, after building I'm seeing a problem of DllRegisterServer failed, which is probably due to the OS binding of the DLLs failing (#284).  I'll leave that with you for now, but let me know if you need me to look at it further.

**Reply by wqweto** [reply]
Yes, the misconfigured `HasPtrSafe` conditional compilation above imported `vbe7.dll` version of `VarPtr` instead of `msvbvm60.dll` one.

Now `regsvr32 UcsFP20TB_win32.dll` works here and there are no missing dependencies showing.

10x for the quick support.
